In [ ]:
!java -version

[0.021s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.021s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
openjdk version "21.0.12.1" 2026-08-18
OpenJDK Runtime Environment (build 21.0.12.1+1-1-24.04.4-Ubuntu)
OpenJDK 64-Bit Server VM (build 21.0.12.1+1-1-24.04.4-Ubuntu, mixed mode, sharing)


In [ ]:

!wget -q --show-progress https://archive.apache.org/dist/hadoop/common/hadoop-3.5.0/hadoop-3.5.0.tar.gz -O /content/hadoop-3.5.0.tar.gz

/content/hadoop-3.5 100%[===================>] 554.35M   115KB/s    in 2h 15m  


In [ ]:

!ls -lh /content/hadoop-3.5.0.tar.gz

ls: cannot access '/content/hadoop-3.5.0.tar.gz': No such file or directory


In [2]:
!apt-get update -qq
!apt-get install -y aria2

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu noble InRelease' does not seem to provide it (sources.list entry misspelt?)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libaria2-0 libcares2
The following NEW packages will be installed:
  aria2 libaria2-0 libcares2
0 upgraded, 3 newly installed, 0 to remove and 14 not upgraded.
Need to get 1,566 kB of archives.
After this operation, 5,692 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu noble/main amd64 libcares2 amd64 1.27.0-1.0ubuntu1 [73.7 kB]
Get:2 http://archive.ubuntu.com/ubuntu noble/universe amd64 libaria2-0 amd64 1.37.0+debian-1build3 [1,105 kB]
Get:3 http://archive.ubuntu.com/ubuntu noble/universe amd64 aria2 amd64 1.37.0+debian-1build3 [387 kB]
Fetched 1,566 kB in 1s (1,314 kB/s)
Selecting previously unselected package libcar

In [3]:
!aria2c -x 16 -s 16 -k 1M \
https://archive.apache.org/dist/hadoop/common/hadoop-3.5.0/hadoop-3.5.0.tar.gz \
-d /content \
-o hadoop-3.5.0.tar.gz


10/07 14:48:38 [NOTICE] Downloading 1 item(s)
 *** Download Progress Summary as of Wed Oct  7 14:49:38 2026 *** 
=
[#1d1031 94MiB/554MiB(16%) CN:16 DL:1.6MiB ETA:4m43s]
FILE: /content/hadoop-3.5.0.tar.gz
-

 *** Download Progress Summary as of Wed Oct  7 14:50:39 2026 *** 
=
[#1d1031 178MiB/554MiB(32%) CN:16 DL:1.4MiB ETA:4m19s]
FILE: /content/hadoop-3.5.0.tar.gz
-

 *** Download Progress Summary as of Wed Oct  7 14:51:39 2026 *** 
=
[#1d1031 260MiB/554MiB(47%) CN:16 DL:1.1MiB ETA:4m9s]
FILE: /content/hadoop-3.5.0.tar.gz
-

 *** Download Progress Summary as of Wed Oct  7 14:52:40 2026 *** 
=
[#1d1031 336MiB/554MiB(60%) CN:16 DL:1.3MiB ETA:2m41s]
FILE: /content/hadoop-3.5.0.tar.gz
-

 *** Download Progress Summary as of Wed Oct  7 14:53:40 2026 *** 
=
[#1d1031 425MiB/554MiB(76%) CN:16 DL:1.5MiB ETA:1m22s]
FILE: /content/hadoop-3.5.0.tar.gz
-

 *** Download Progress Summary as of Wed Oct  7 14:54:41 2026 *** 
=
[#1d1031 512MiB/554MiB(92%) CN:16 DL:1.2MiB ETA:32s]
FILE: /content/hadoop-3

In [4]:
!ls -lh /content/hadoop-3.5.0.tar.gz

-rw-r--r-- 1 root root 555M Oct  7 14:55 /content/hadoop-3.5.0.tar.gz


In [5]:
!tar -xzf /content/hadoop-3.5.0.tar.gz -C /content

In [6]:
!ls /content/hadoop-3.5.0

bin	    etc      libexec	      LICENSE.txt    README.txt
compose     include  LICENSE-binary   NOTICE-binary  sbin
Dockerfile  lib      licenses-binary  NOTICE.txt     share


In [8]:
import os
import subprocess

java_home = subprocess.check_output(
    "dirname $(dirname $(readlink -f $(which java)))",
    shell=True,
    text=True
).strip()

os.environ["JAVA_HOME"] = java_home
os.environ["HADOOP_HOME"] = "/content/hadoop-3.5.0"
os.environ["PATH"] += ":/content/hadoop-3.5.0/bin:/content/hadoop-3.5.0/sbin"

print("JAVA_HOME =", os.environ["JAVA_HOME"])
print("HADOOP_HOME =", os.environ["HADOOP_HOME"])

JAVA_HOME = /usr/lib/jvm/java-21-openjdk-amd64
HADOOP_HOME = /content/hadoop-3.5.0


In [9]:
import os

hadoop_env = "/content/hadoop-3.5.0/etc/hadoop/hadoop-env.sh"

with open(hadoop_env, "a") as f:
    f.write(f"\nexport JAVA_HOME={os.environ['JAVA_HOME']}\n")

print("Đã cập nhật:", hadoop_env)

Đã cập nhật: /content/hadoop-3.5.0/etc/hadoop/hadoop-env.sh


In [10]:
!hadoop version

[0.036s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.037s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
Hadoop 3.5.0
Source code repository https://github.com/apache/hadoop.git -r dbcc7cd797100e6b32cd84f85b53a5193a5f9af0
Compiled by cnauroth on 2026-03-24T16:56Z
Compiled on platform linux-x86_64
Compiled with protoc 3.25.5
From source with checksum 64582ed0a62a169c39a19c734a8f33b5
This command was run using /content/hadoop-3.5.0/share/hadoop/common/hadoop-common-3.5.0.jar


In [11]:
!mkdir -p /content/hadoop_data/namenode
!mkdir -p /content/hadoop_data/datanode
!ls -ld /content/hadoop_data/*

drwxr-xr-x 2 root root 4096 Oct  7 15:16 /content/hadoop_data/datanode
drwxr-xr-x 2 root root 4096 Oct  7 15:16 /content/hadoop_data/namenode


In [13]:
%%writefile /content/hadoop-3.5.0/etc/hadoop/core-site.xml
<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
<configuration>
    <property>
        <name>fs.defaultFS</name>
        <value>hdfs://localhost:9000</value>
    </property>
</configuration>

Overwriting /content/hadoop-3.5.0/etc/hadoop/core-site.xml


In [14]:
%%writefile /content/hadoop-3.5.0/etc/hadoop/hdfs-site.xml
<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
<configuration>
    <property>
        <name>dfs.replication</name>
        <value>1</value>
    </property>

    <property>
        <name>dfs.namenode.name.dir</name>
        <value>file:///content/hadoop_data/namenode</value>
    </property>

    <property>
        <name>dfs.datanode.data.dir</name>
        <value>file:///content/hadoop_data/datanode</value>
    </property>
</configuration>

Overwriting /content/hadoop-3.5.0/etc/hadoop/hdfs-site.xml


In [15]:
%%writefile /content/hadoop-3.5.0/etc/hadoop/mapred-site.xml
<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
<configuration>
    <property>
        <name>mapreduce.framework.name</name>
        <value>local</value>
    </property>
</configuration>

Overwriting /content/hadoop-3.5.0/etc/hadoop/mapred-site.xml


In [16]:
!hadoop getconf -confKey fs.defaultFS
!hadoop getconf -confKey dfs.replication
!hadoop getconf -confKey mapreduce.framework.name

ERROR: getconf is not COMMAND nor fully qualified CLASSNAME.
Usage: hadoop [OPTIONS] SUBCOMMAND [SUBCOMMAND OPTIONS]
 or    hadoop [OPTIONS] CLASSNAME [CLASSNAME OPTIONS]
  where CLASSNAME is a user-provided Java class

  OPTIONS is none or any of:

buildpaths                       attempt to add class files from build tree
--config dir                     Hadoop config directory
--debug                          turn on shell script debug mode
--help                           usage information
hostnames list[,of,host,names]   hosts to use in worker mode
hosts filename                   list of hosts to use in worker mode
loglevel level                   set the log4j level for this command
workers                          turn on worker mode

  SUBCOMMAND is one of:


    Admin Commands:

daemonlog     get/set the log level for each daemon

    Client Commands:

archive       create a Hadoop archive
checknative   check native Hadoop and compression libraries availability
classpath     

In [17]:
!hdfs getconf -confKey fs.defaultFS
!hdfs getconf -confKey dfs.replication

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
hdfs://localhost:9000
[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
1


In [18]:
!cat /content/hadoop-3.5.0/etc/hadoop/mapred-site.xml

<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
<configuration>
    <property>
        <name>mapreduce.framework.name</name>
        <value>local</value>
    </property>
</configuration>


In [20]:
import xml.etree.ElementTree as ET

tree = ET.parse("/content/hadoop-3.5.0/etc/hadoop/mapred-site.xml")
root = tree.getroot()

for prop in root.findall("property"):
    if prop.find("name").text == "mapreduce.framework.name":
        print(prop.find("value").text)

local


In [21]:
!hdfs getconf -confKey fs.defaultFS
!hdfs getconf -confKey dfs.replication

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
hdfs://localhost:9000
[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
1


In [22]:
!cat /content/hadoop-3.5.0/etc/hadoop/mapred-site.xml

<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
<configuration>
    <property>
        <name>mapreduce.framework.name</name>
        <value>local</value>
    </property>
</configuration>


In [23]:
!hdfs namenode -format -force

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
2026-10-07 15:25:58,604 INFO namenode.NameNode: STARTUP_MSG: 
/************************************************************
STARTUP_MSG: Starting NameNode
STARTUP_MSG:   host = d7a607c39a16/172.28.0.12
STARTUP_MSG:   args = [-format, -force]
STARTUP_MSG:   version = 3.5.0
STARTUP_MSG:   classpath = /content/hadoop-3.5.0/etc/hadoop:/content/hadoop-3.5.0/share/hadoop/common/lib/kerb-crypto-2.0.3.jar:/content/hadoop-3.5.0/share/hadoop/common/lib/commons-math3-3.6.1.jar:/content/hadoop-3.5.0/share/hadoop/common/lib/guava-33.4.8-jre.jar:/content/hadoop-3.5.0/share/hadoop/common/lib/jetty-util-ajax-9.4.58.v20250814.jar:/content/hadoop-3.5.0/share/hadoop/common/lib/nett

In [24]:
!hdfs --daemon start namenode

In [25]:
!hdfs --daemon start datanode

In [26]:
!jps

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
14114 Jps
14003 DataNode
13855 NameNode


In [27]:
!hdfs dfsadmin -report

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
Configured Capacity: 115658190848 (107.72 GB)
Present Capacity: 91506655232 (85.22 GB)
DFS Remaining: 91506630656 (85.22 GB)
DFS Used: 24576 (24 KB)
DFS Used%: 0.00%
Replicated Blocks:
	Under replicated blocks: 0
	Blocks with corrupt replicas: 0
	Missing blocks: 0
	Missing blocks (with replication factor 1): 0
	Badly Distributed Blocks: 0
	Low redundancy blocks with highest priority to recover: 0
	Pending deletion blocks: 0
Erasure Coded Block Groups: 
	Low redundancy block groups: 0
	Block groups with corrupt internal blocks: 0
	Missing block groups: 0
	Badly Distributed Blocks: 0
	Low redundancy blocks with highest priority to recover: 0
	Pending deletion block

In [28]:
!hdfs dfs -ls /

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate


In [29]:
from google.colab import files
uploaded = files.upload()

Saving orders_2026_09.csv to orders_2026_09.csv


In [30]:
!ls -lh /content/orders_2026_09.csv

-rw-r--r-- 1 root root 117K Oct  7 15:30 /content/orders_2026_09.csv


In [31]:
!head -5 /content/orders_2026_09.csv

transaction_id,order_date,order_time,province,category,channel,quantity,unit_price,amount,status,shipping_days
TX000001,9/30/2026,2:25:00,Ha Noi,Beauty,web,1,294947,294947,SUCCESS,2
TX000002,9/23/2026,23:58:00,Da Nang,Grocery,mobile,1,265514,265514,SUCCESS,3
TX000003,9/5/2026,15:38:00,Ha Noi,Electronics,partner,1,12741238,12741238,SUCCESS,3
TX000004,9/9/2026,18:01:00,Hai Phong,Electronics,mobile,1,8846997,8846997,SUCCESS,4


In [32]:
!hdfs dfs -mkdir -p /user/student/ecommerce/input
!hdfs dfs -mkdir -p /user/student/ecommerce/output

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate


In [33]:
!hdfs dfs -ls -R /user/student/ecommerce

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
drwxr-xr-x   - root supergroup          0 2026-10-07 15:31 /user/student/ecommerce/input
drwxr-xr-x   - root supergroup          0 2026-10-07 15:31 /user/student/ecommerce/output


In [34]:
!hdfs dfs -put -f /content/orders_2026_09.csv /user/student/ecommerce/input/

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate


In [35]:
!hdfs dfs -ls -h /user/student/ecommerce/input

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
Found 1 items
-rw-r--r--   1 root supergroup    116.1 K 2026-10-07 15:31 /user/student/ecommerce/input/orders_2026_09.csv


In [36]:
!hdfs dfs -head /user/student/ecommerce/input/orders_2026_09.csv

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
transaction_id,order_date,order_time,province,category,channel,quantity,unit_price,amount,status,shipping_days
TX000001,9/30/2026,2:25:00,Ha Noi,Beauty,web,1,294947,294947,SUCCESS,2
TX000002,9/23/2026,23:58:00,Da Nang,Grocery,mobile,1,265514,265514,SUCCESS,3
TX000003,9/5/2026,15:38:00,Ha Noi,Electronics,partner,1,12741238,12741238,SUCCESS,3
TX000004,9/9/2026,18:01:00,Hai Phong,Electronics,mobile,1,8846997,8846997,SUCCESS,4
TX000005,9/10/2026,12:50:00,Hai Phong,Beauty,web,1,593405,593405,RETURNED,6
TX000006,9/11/2026,10:10:00,Dong Nai,Home,web,1,2839709,2839709,FAILED,5
TX000007,9/13/2026,1:19:00,Ha Noi,Beauty,web,1,839161,839161,SUCCESS,1
TX000008,9/29/2026,13:20

In [37]:
%%writefile /content/mapper_revenue.py
import sys
import csv

reader = csv.DictReader(sys.stdin)

for row in reader:
    if row["status"] == "SUCCESS":
        category = row["category"]
        amount = int(row["amount"])
        print(f"{category}\t{amount}")

Writing /content/mapper_revenue.py


In [38]:
%%writefile /content/reducer_revenue.py
import sys

current_category = None
total = 0

for line in sys.stdin:
    line = line.strip()

    if not line:
        continue

    category, amount = line.split("\t")
    amount = int(amount)

    if current_category == category:
        total += amount
    else:
        if current_category is not None:
            print(f"{current_category}\t{total}")

        current_category = category
        total = amount

if current_category is not None:
    print(f"{current_category}\t{total}")

Writing /content/reducer_revenue.py


In [39]:
!ls -lh /content/mapper_revenue.py /content/reducer_revenue.py

-rw-r--r-- 1 root root 223 Oct  7 15:35 /content/mapper_revenue.py
-rw-r--r-- 1 root root 494 Oct  7 15:35 /content/reducer_revenue.py


In [40]:
!cat /content/orders_2026_09.csv | \
python3 /content/mapper_revenue.py | \
sort | \
python3 /content/reducer_revenue.py

Beauty	111032003
Book	39306022
Electronics	3703769311
Fashion	218597763
Grocery	80983147
Home	410315065
Sport	166194354
Toy	83881779


In [41]:
!cat /content/orders_2026_09.csv | \
python3 /content/mapper_revenue.py | \
sort | \
python3 /content/reducer_revenue.py

Beauty	111032003
Book	39306022
Electronics	3703769311
Fashion	218597763
Grocery	80983147
Home	410315065
Sport	166194354
Toy	83881779


In [42]:
!cat /content/orders_2026_09.csv | \
python3 /content/mapper_revenue.py | \
sort | \
python3 /content/reducer_revenue.py

Beauty	111032003
Book	39306022
Electronics	3703769311
Fashion	218597763
Grocery	80983147
Home	410315065
Sport	166194354
Toy	83881779


In [43]:
!hdfs dfs -rm -r -f /user/student/ecommerce/output/revenue_by_category

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate


In [44]:
!hadoop jar /content/hadoop-3.5.0/share/hadoop/tools/lib/hadoop-streaming*.jar \
-files /content/mapper_revenue.py,/content/reducer_revenue.py \
-mapper "python3 mapper_revenue.py" \
-reducer "python3 reducer_revenue.py" \
-input /user/student/ecommerce/input/orders_2026_09.csv \
-output /user/student/ecommerce/output/revenue_by_category

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
2026-10-07 15:38:59,626 WARN impl.MetricsSystemImpl: JobTracker metrics system already initialized!
2026-10-07 15:39:00,099 INFO mapred.FileInputFormat: Total input files to process : 1
2026-10-07 15:39:00,188 INFO mapreduce.JobSubmitter: number of splits:1
2026-10-07 15:39:00,564 INFO mapreduce.JobSubmitter: Submitting tokens for job: job_local597866926_0001
2026-10-07 15:39:00,566 INFO mapreduce.JobSubmitter: Executing with tokens: []
2026-10-07 15:39:00,968 INFO mapred.LocalDistributedCacheManager: Localized file:/content/mapper_revenue.py as file:/tmp/hadoop-root/mapred/local/job_local597866926_0001_03e05a6a-900b-41d6-9f91-f5b6d7bfca24/mapper_revenue.py
2026-

In [45]:
!hdfs dfs -ls /user/student/ecommerce/output/revenue_by_category

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
Found 2 items
-rw-r--r--   1 root supergroup          0 2026-10-07 15:39 /user/student/ecommerce/output/revenue_by_category/_SUCCESS
-rw-r--r--   1 root supergroup        133 2026-10-07 15:39 /user/student/ecommerce/output/revenue_by_category/part-00000


In [46]:
!hdfs dfs -cat /user/student/ecommerce/output/revenue_by_category/part-*

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
Beauty	111032003
Book	39306022
Electronics	3703769311
Fashion	218597763
Grocery	80983147
Home	410315065
Sport	166194354
Toy	83881779


In [47]:
!hdfs dfs -cat /user/student/ecommerce/output/revenue_by_category/part-* | sort -k2,2nr | head -3

Electronics	3703769311
Home	410315065
Fashion	218597763


In [48]:
!rm -f /content/result_revenue_by_category.txt

!hdfs dfs -get \
/user/student/ecommerce/output/revenue_by_category/part-00000 \
/content/result_revenue_by_category.txt

[0.001s][warning][os,container] Cgroup memory controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate
[0.001s][warning][os,container] Cgroup cpu controller path at '/sys/fs/cgroup' seems to have moved to '/../../jupyter-children', detected limits won't be accurate


In [49]:
!cat /content/result_revenue_by_category.txt

Beauty	111032003
Book	39306022
Electronics	3703769311
Fashion	218597763
Grocery	80983147
Home	410315065
Sport	166194354
Toy	83881779


In [50]:
!sort -k2,2nr /content/result_revenue_by_category.txt | \
head -3 > /content/top3_revenue.txt

In [51]:
!cat /content/top3_revenue.txt

Electronics	3703769311
Home	410315065
Fashion	218597763


In [52]:
from google.colab import files

files.download('/content/result_revenue_by_category.txt')
files.download('/content/top3_revenue.txt')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [53]:
from google.colab import files

files.download('/content/result_revenue_by_category.txt')
files.download('/content/top3_revenue.txt')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## BÁO CÁO KẾT QUẢ THỰC HÀNH HADOOP

### 1. Kết quả doanh thu theo nhóm sản phẩm

| Hạng | Category | Tổng doanh thu |
|---|---|---:|
| 1 | Electronics | 3,703,769,311 |
| 2 | Home | 410,315,065 |
| 3 | Fashion | 218,597,763 |

### 2. Nhận xét kết quả

Dữ liệu đầu vào là file `orders_2026_09.csv`, chứa thông tin các giao dịch thương mại điện tử trong tháng 09/2026.  
Bài toán chỉ tính doanh thu đối với những giao dịch có trạng thái `SUCCESS`.  
Mapper đọc từng dòng CSV và phát sinh cặp key-value theo dạng `category → amount`.  
Sau bước Shuffle/Sort, các bản ghi có cùng category được gom lại với nhau.  
Reducer cộng tổng amount của từng category và ghi kết quả vào file `part-00000`.  
Kết quả cho thấy Electronics có doanh thu cao nhất với 3,703,769,311, tiếp theo là Home với 410,315,065 và Fashion với 218,597,763.  
Kết quả này có thể hỗ trợ doanh nghiệp ưu tiên quản lý tồn kho, chương trình khuyến mại và nguồn lực giao hàng cho các nhóm sản phẩm có doanh thu cao.

### 3. Hạn chế của xử lý batch

Hadoop MapReduce phù hợp với xử lý dữ liệu lớn theo lô, tuy nhiên kết quả không được cập nhật tức thời. Khi dữ liệu mới phát sinh, cần chạy lại job để tạo kết quả mới, vì vậy mô hình này không phù hợp với các bài toán yêu cầu phản hồi thời gian thực.